src/main/scala/com.snowplowanalytics.sauna/utils/Command.scala

```diff
@@ -0,0 +1,215 @@
+  )
+  implicit val commandEnvelopeReads: Reads[CommandEnvelope] = Json.reads[CommandEnvelope]
+
+  /**
+   * Attempts to extract a Sauna command from a [[JsValue]]. If successful,
+   * passes it on to the processing method.
+   *
+   * @param json The [[JsValue]] to extract a command from.
+   * @tparam T The type of the command's data.
+   * @return Right containing a tuple of the command's envelope and data
+   *         if the extraction and processing was successful, Left containing
+   *         an error message otherwise.
+   */
+  def extractCommand[T](json: JsValue)(implicit tReads: Reads[T]): Either[String, (CommandEnvelope, T)] = {
+    json.validate[SelfDescribing] match {
+      case JsSuccess(selfDescribing, _) =>
+        validateSelfDescribing(selfDescribing) match {
+          case None =>
+            selfDescribing.data.validate[SaunaCommand] match {
+              case JsSuccess(command, _) => processCommand[T](command)
+              case JsError(error) => Left(s"Encountered an issue while parsing Sauna command: $error")
+            }
+          case Some(error) => Left(s"Could not validate command JSON: $error")
+        }
+      case JsError(error) => Left(s"Encountered an issue while parsing self-describing JSON: $error")
+    }
+  }
+
+  /**
+   * Processes a Sauna command, validating its' envelope and extracting the data.
+   *
+   * @param command A Sauna commmand.
+   * @tparam T The type of the command's data.
+   * @return Right containing a tuple of the command's envelope and data
+   *         if the processing was successful, Left containing
+   *         an error message otherwise.
+   */
+  def processCommand[T](command: SaunaCommand)(implicit tReads: Reads[T]): Either[String, (CommandEnvelope, T)] = {
+    validateSelfDescribing(command.envelope) match {
+      case None =>
+        command.envelope.data.validate[CommandEnvelope] match {
+          case JsSuccess(envelope, _) =>
+            validateSelfDescribing(command.command) match {
+              case None =>
+                command.command.data.validate[T] match {
+                  case JsSuccess(data, _) => Right((envelope, data))
+                  case JsError(error) => Left(s"Encountered an issue while parsing Sauna command data: $error")
+                }
+              case Some(error) => Left(s"Could not validate command data JSON: $error")
+            }
+          case JsError(error) => Left(s"Encountered an issue while parsing Sauna command envelope: $error")
+        }
+      case Some(error) => Left(s"Could not validate command envelope JSON: ${error}")
+    }
+  }
+
+  /**
+   * Validates self-describing data against its' schema.
+   *
+   * @param selfDescribing The self-describing data to validate.
+   * @return None if the data was successfully validated against the schema,
+   *         Some with an error message otherwise.
+   */
+  def validateSelfDescribing(selfDescribing: SelfDescribing): Option[String] = {
```

Comment:
The reason this is commented out is so that integration tests don't throw "schema not available on iglu" errors. I'll uncomment it when all of the new specs are on iglu-central.